### PR TITLE
[REEF-663] Make Http Endpoint available for client after submit

### PR DESCRIPTION
**Pull request by jwang98052**
-This PR is to return HttpEndPoint to the client after Job is submitted
-At Java side, the http endpoint is written to JobSubmission folder in HDFS and Driver folder
-At REEFClient side, HttpClientHelper is added to pull the http endpoint from the file and return it in Submit()
-E2e test cases are updated to get the URI and verify on it.
-The new Submit API is noted as [Ustable] as we might change it in future
-The change is part of the change in tlc branch with some modifications

JIRA: [REEF-663](https://issues.apache.org/jira/browse/REEF-663)

This closes #

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Client/Common/JavaClientLauncher.cs [hunk 1]
```diff
@@ -100,6 +100,7 @@ private string AssembleArguments(string javaClassName, params string[] parameter
             IList<string> arguments = new List<string>();
             arguments.Add("-cp");
             arguments.Add(GetClientClasspath());
+            arguments.Add("-Djava.util.logging.config.class=org.apache.reef.util.logging.Config");
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/Common/JavaClientLauncher.cs]: This seems orthogonal to the goals of the PR. Comment?

File: lang/cs/Org.Apache.REEF.Common/Files/REEFFileNames.cs [hunk 2]
```diff
@@ -229,6 +230,11 @@ public string GetBridgeExeConfigPath()
             return Path.Combine(REEF_BASE_FOLDER, BRIDGE_EXE_CONFIG_NAME);
         }
 
+        /// <summary>
+        /// </summary>
+        /// <returns>File name that contains the dfs path for the DriverHttpEndpoint</returns>
+        public string DriverHttpEndpoint { get { return DRIVER_HTTP_ENDPOINT_FILE_NAME; } }
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Files/REEFFileNames.cs]: This needs to have an `[Unstable]` annotation, as it is likely to go away once we solve this for good.

**Comment by markusweimer** [comment]
I've done a pass. Two overarching comments:

  * `SubmitAsync` on a method that isn't `async` is odd. Also, it grabs the name such that we can't use it in future APIs. Maybe use `SubmitAndGetDriverUrl`?
  * I don't think we should change any of the examples or applications in the code base to use this API. It is temporary anyhow.

File: lang/cs/Org.Apache.REEF.Client/Common/JavaClientLauncher.cs (hunk 1 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/Common/JavaClientLauncher.cs]: I can remove for now.

File: lang/cs/Org.Apache.REEF.Common/Files/REEFFileNames.cs (hunk 2 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Files/REEFFileNames.cs]: Done

**Comment by jwang98052** [comment]
In the new Submit method in LocalClient, we do Launch Java with a separate  Task.Run(() =>. And then pull the result back. I have renamed it anyway as it is temporary. 
I have changed most places to use old Submit. Keep one test case to call the new  SubmitAndGetDriverUrl because I would like to have a test case for it.

**Comment by jwang98052** [comment]
The change has pushed in.

File: lang/cs/Org.Apache.REEF.Client/Common/HttpClientHelper.cs [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+                    break;
+                }
+
+                Thread.Sleep(MilliSecondsToWaitBeforeNextConnectAttempt);
+            }
+
+            return null;
+        }
+
+        private static bool ShouldRetry(HttpStatusCode httpStatusCode)
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Client/Common/HttpClientHelper.cs]: Is this function used? Looks like only the other version is used

File: lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs [hunk 4]
```diff
@@ -51,6 +51,7 @@ public class ReefFunctionalTest
 
         private const string Local = "local";
         private const string YARN = "yarn";
+        private const int SleppTime = 1000;
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs]: Typo: SleepTime

File: lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs [hunk 5]
```diff
@@ -139,14 +140,35 @@ protected void ValidateSuccessForLocalRuntime(int numberOfEvaluatorsToClose, str
...
+                string[] failedTaskIndicators = lines.Where(s => s.Contains(failedTaskIndication)).ToArray();
+                string[] failedIndicators = lines.Where(s => s.Contains(failedEvaluatorIndication)).ToArray();
+                Assert.IsTrue(successIndicators.Count() == numberOfEvaluatorsToClose);
+                Assert.IsFalse(failedTaskIndicators.Any());
+                Assert.IsFalse(failedIndicators.Any());
+            }
+            else
+            {
+                Console.WriteLine("Cannot read from log file");
+                Assert.IsNotNull(null);
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs]: Could be Assert.Fail() (which makes the intention a bit more evident)

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java [hunk 6]
```diff
@@ -43,8 +44,8 @@
 import org.apache.reef.tang.exceptions.InjectionException;
 import org.apache.reef.tang.formats.ConfigurationSerializer;
 import org.apache.reef.util.JARFileMaker;
-
 import javax.inject.Inject;
+import java.io.*;
```
Comment (tcNickolas) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java]: If we're importing `io.*` we can remove subsequent imports of individual classes. But from our code it looks like we're trying not to import * but rather list individual classes we need.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java [hunk 7]
```diff
@@ -207,7 +206,65 @@ private void launch(final YarnSubmissionFromCS yarnSubmission) throws IOExceptio
...
+    final FileSystem fs = FileSystem.get(yarnConfiguration);
+    final Path httpEndpointPath = new Path(dfsPath, fileNames.getDriverHttpEndpoint());
+
+    String trackingUri = null;
+    for (int i = 0; i < 60; i++) {
+      try {
+        LOG.log(Level.INFO, "Attempt " + i + " reading " + httpEndpointPath.toString());
+        if (fs.exists(httpEndpointPath)) {
+          FSDataInputStream input = fs.open(httpEndpointPath);
+          BufferedReader reader = new BufferedReader(new InputStreamReader(input));
```
Comment (tcNickolas) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java]: In REEF-768 we want to eliminate high-priority warnings from Findbugs; this line will (likely) produce a new warning, `DM_DEFAULT_ENCODING: Reliance on default encoding` which complains about not setting encoding explicitly. Adding "UTF-8" as second parameter for `InputStreamReader` constructor should fix it.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java [hunk 8]
```diff
@@ -207,7 +206,65 @@ private void launch(final YarnSubmissionFromCS yarnSubmission) throws IOExceptio
...
+      }
+    }
+
+    if (null == trackingUri) {
+      trackingUri = "";
+      LOG.log(Level.WARNING, "Failed reading " + httpEndpointPath.toString());
+    }
+
+    final File driverHttpEndpointFile = new File(driverFolder, fileNames.getDriverHttpEndpoint());
+    BufferedWriter out = new BufferedWriter(new FileWriter(driverHttpEndpointFile));
```
Comment (tcNickolas) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java]: Same Findbugs warning here; `new OutputStreamWriter(file, "UTF-8")` instead of `FileWriter` should fix it

File: lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/generic/JobDriver.java [hunk 9]
```diff
@@ -188,6 +198,17 @@ private void setupBridge(final ClrHandlersInitializer initializer) {
       }
 
       final String portNumber = httpServer == null ? null : Integer.toString((httpServer.getPort()));
+      if (portNumber != null){
+        try {
+          final File outputFileName = new File(reefFileNames.getDriverHttpEndpoint());
+          BufferedWriter out = new BufferedWriter(new FileWriter(outputFileName));
```
Comment (tcNickolas) [inline lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/generic/JobDriver.java]: Same warning

**Comment by tcNickolas** [comment]
I've done a pass

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java (hunk 6 above)
Comment (jwang98052) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java]: Done

File: lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs (hunk 5 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs]: Done

File: lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs (hunk 4 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs]: Done

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java (hunk 7 above)
Comment (jwang98052) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java]: Done

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java (hunk 8 above)
Comment (jwang98052) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java]: OutputStreamWriter doesn't accept File as a parameter.

**Comment by jwang98052** [comment]
Review comments addressed and changes pushed in.

**Comment by markusweimer** [comment]
I'll test and merge.
